File: frontend/src/api/runs.ts
/**
 * 运行记录 API 封装（specs/011，契约主定义 specs/011-context-compression-run-records/contracts/runs-api.md）。
 * 类型 MUST 与后端 schemas/runs.py 逐字段对齐，变更先改契约。
 */
import { http } from './request'

// ---- 枚举与常量（契约唯一主定义，实现侧只消费）----

export type RunStatus = 'running' | 'succeeded' | 'partial' | 'failed' | 'cancelled'

export const RUN_STATUSES: RunStatus[] = ['running', 'succeeded', 'partial', 'failed', 'cancelled']

export const RUN_STATUS_META: Record<RunStatus, { text: string; color: string }> = {
  running: { text: '运行中', color: 'processing' },
  succeeded: { text: '成功', color: 'success' },
  partial: { text: '部分完成', color: 'warning' },
  failed: { text: '失败', color: 'error' },
  cancelled: { text: '已取消', color: 'default' },
}

// ---- 数据结构 ----

export interface RunSummary {
  run_id: string
  conversation_id: number | null
  agent_name: string
  model_name: string
  status: RunStatus
  end_reason: string
  error_summary: string | null
  started_at: string
  finished_at: string | null
  total_duration_ms: number | null
  model_call_count: number
  tool_call_count: number
  prompt_tokens: number | null
  completion_tokens: number | null
  total_tokens: number | null
  first_output_ms: number | null
}

export interface RunListResponse {
  items: RunSummary[]
  total: number
  page: number
  page_size: number
}

export interface RunEventOut {
  seq: number
  event_type: string
  round: number | null
  call_id: string | null
  data: Record<string, unknown>
  created_at: string
}

export interface RunDetail {
  summary: RunSummary
  events: RunEventOut[]
}

export interface RunPayloadMeta {
  id: number
  call_id: string
  payload_type: string
  char_count: number
}

export interface RunPayloadContent extends RunPayloadMeta {
  content: string
}

export const PAYLOAD_TYPE_TEXT: Record<string, string> = {
  model_input: '模型输入',
  model_output: '模型输出',
  tool_params: '工具参数',
  tool_result: '工具结果',
  compression_input: '压缩输入',
  compression_output: '压缩摘要',
}

export interface RunListQuery {
  status?: RunStatus
  agent_id?: number
  conversation_id?: number
  page?: number
  page_size?: number
}

// ---- API ----


// ---- 会话消息过程回放（011 优化①：刷新后回显工具调用全过程）----

export type ReplayItem =
  | { kind: 'reasoning' | 'content'; round: number | null; text: string }
  | {
      kind: 'tool'
      round: number | null
      callId: string
      toolName: string
      displayName: string
      toolType: 'builtin' | 'mcp' | 'skill'
      serverName: string | null
      status: 'success' | 'error' | 'denied' | 'cancelled'
      durationMs: number | null
      paramsSummary: string
      resultSummary: string
      paramsText: string
      resultText: string
    }

export interface RunReplay {
  run_id: string
  reply_message_id: number | null
  status: RunStatus
  items: ReplayItem[]
}

export const runsApi = {
  list(query: RunListQuery = {}): Promise<RunListResponse> {
    const params = new URLSearchParams()
    if (query.status) params.set('status', query.status)
    if (query.agent_id != null) params.set('agent_id', String(query.agent_id))
    if (query.conversation_id != null) params.set('conversation_id', String(query.conversation_id))
    params.set('page', String(query.page ?? 1))
    params.set('page_size', String(query.page_size ?? 20))
    return http.get(`/runs?${params.toString()}`)
  },
  detail(runId: string): Promise<RunDetail> {
    return http.get(`/runs/${runId}`)
  },
  payloads(runId: string): Promise<RunPayloadMeta[]> {
    return http.get(`/runs/${runId}/payloads`)
  },
  payload(runId: string, payloadId: number): Promise<RunPayloadContent> {
    return http.get(`/runs/${runId}/payloads/${payloadId}`)
  },
  conversationRuns(conversationId: number): Promise<RunSummary[]> {
    return http.get(`/conversations/${conversationId}/runs`)
  },
  conversationReplays(conversationId: number): Promise<RunReplay[]> {
    return http.get(`/conversations/${conversationId}/run-replays`)
  },
}

File: backend/app/schemas/runs.py
"""运行记录 API 契约实现（specs/011，契约主定义 contracts/runs-api.md）。

字段与枚举 MUST 与契约逐条对齐，变更先改契约。
"""

from typing import Literal

from pydantic import BaseModel, Field

# ---- 枚举与常量（契约唯一主定义）----

RunStatusLiteral = Literal["running", "succeeded", "partial", "failed", "cancelled"]

PAGE_DEFAULT = 1
PAGE_SIZE_DEFAULT = 20
PAGE_SIZE_MAX = 100


class RunSummary(BaseModel):
    """运行摘要（列表项；全部来自 runs 行快照，无关联查询）。"""

    run_id: str
    conversation_id: int | None
    agent_name: str
    model_name: str
    status: RunStatusLiteral
    # ---- 014 增补：运行启动时快照的会话工作空间（None = 未选择，data-model §1.2）----
    workspace_path: str | None = None
    end_reason: str = ""
    error_summary: str | None = None
    started_at: str
    finished_at: str | None = None
    total_duration_ms: int | None = None
    model_call_count: int = 0
    tool_call_count: int = 0
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    total_tokens: int | None = None
    first_output_ms: int | None = None


class RunListResponse(BaseModel):
    """GET /api/runs 响应。"""

    items: list[RunSummary]
    total: int
    page: int
    page_size: int


class RunEventOut(BaseModel):
    """详情时间线事件（安全负载，不含任何透传字段/正文全文/增量）。"""

    seq: int
    event_type: str
    round: int | None = None
    call_id: str | None = None
    data: dict
    created_at: str


class RunDetailResponse(BaseModel):
    """GET /api/runs/{run_id} 响应。"""

    summary: RunSummary
    events: list[RunEventOut]


class RunPayloadMeta(BaseModel):
    """载荷元数据（不含 content）。"""

    id: int
    call_id: str
    payload_type: str
    char_count: int


class RunPayloadContent(BaseModel):
    """GET 单条载荷响应（脱敏后完整内容）。"""

    id: int
    call_id: str
    payload_type: str
    char_count: int
    content: str
